File: StatelessNFJAVA/src/main/java/NAT.java

public class NAT extends NF {
    private RC nowRC;
    private String ClusterID;
    public NAT(RC newRC, String newClusterID)
    {
        nowRC = newRC;
        ClusterID = newClusterID;
    }
    private String select_IP_Port(String listIPPort, String tupleof5)
    {
        return "127.0.0.1:2131";
    }
    private String update(String listIPPort, String IPport)
    {
        return listIPPort;
    }
    private Packet updatePacketHeader(Packet P, String IPport)
    {
        return P;
    }
    public void ProcessPacket(Packet P) {
        String tupleof5 = P.Get("5-tuple");
        String PIPport = P.Get("IPport");
        String IPport = nowRC.ReadRC(tupleof5);
        if (IPport == null)
        {
            String listIPPort = nowRC.ReadRC(ClusterID);
            IPport = select_IP_Port(listIPPort, tupleof5);
            listIPPort = update(listIPPort, IPport);
            nowRC.WriteRC(ClusterID, listIPPort);
            nowRC.WriteRC(tupleof5, IPport);
            String reversetupleof5 = tupleof5;
            nowRC.WriteRC(reversetupleof5, PIPport);
        }
        Packet P1 = updatePacketHeader(P, IPport);
        PacketSender.sendPacket(P1);
    }
}

File: StatelessNFC/NFS/nat.cpp
#include "nat.h"

NAT::NAT(RC* rc, char* ClusterID, uint16_t lClusterID)
{
    this->rc = rc;
    this->ClusterID = (char*)malloc(lClusterID);
    memcpy(this->ClusterID, ClusterID, lClusterID);
    this->lClusterID = lClusterID;
}
void NAT::ProcessPacket(char *P, uint16_t Packetlen)
{
    char tupleof5[20]; uint16_t ltupleof5;
    char pipport[8]; uint16_t lpipport;
    char ipport[8]; uint16_t lipport;
    rc->readRC(tupleof5,ltupleof5,ipport,lipport);
    if (lipport == 0)
    {
        char ipportlist[100]; uint16_t lipportlist;
        char newipportlist[100]; uint16_t lnewipportlist;
        rc->readRC(tupleof5,ltupleof5,ipportlist,lipportlist);
        select_IP_Port(ipportlist,lipportlist,tupleof5,ltupleof5,ipport,lipport);
        update(ipportlist,lipportlist,ipport,lipport,newipportlist,lnewipportlist);
        rc->writeRC(ClusterID,lClusterID,ipportlist,lipportlist);
        rc->writeRC(tupleof5,ltupleof5,ipport,lipport);
        char* reversetupleof5 = tupleof5; int lreversetupleof5 = ltupleof5;
        rc->writeRC(reversetupleof5,lreversetupleof5,pipport,lpipport);
    }
    char* p1 = (char*)malloc(MAX_PACKET_LEN); uint16_t lp1;
    updatePacketHeader(P,Packetlen,ipport,lipport,p1,lp1);
    //sendPacket(p1,lp1);
}
void NAT::select_IP_Port(char *ipportlist, uint16_t l1, char *tupleof5, uint16_t l2, char *dest, uint16_t &ldest)
{
    ldest=8;
}
void NAT::update(char *ipportlist, uint16_t l1, char *ipport, uint16_t l2, char *dest, uint16_t &ldest)
{
    ldest=l1;
}
void NAT::updatePacketHeader(char *P, uint16_t lP, char *ipport, uint16_t l1, char *dest, uint16_t &ldest)
{
    ldest=lP;
}
